tracer/src/Datadog.Trace/Util/AsyncManualResetEvent.cs

```diff
@@ -0,0 +1,121 @@
+// <copyright file="AsyncManualResetEvent.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+#nullable enable
+
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Datadog.Trace.Util;
+
+// Based on: https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Coordination/AsyncManualResetEvent.cs
+
+internal class AsyncManualResetEvent
+{
+    private readonly object _mutex;
+    private TaskCompletionSource<object?> _tcs;
+
+    public AsyncManualResetEvent(bool set)
+    {
+        _mutex = new object();
+        _tcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+        if (set)
+        {
+            _tcs.TrySetResult(null);
+        }
+    }
+
+    public AsyncManualResetEvent()
+        : this(false)
+    {
+    }
+
+    public bool IsSet
+    {
+        get
+        {
+            lock (_mutex)
+            {
+                return _tcs.Task.IsCompleted;
+            }
+        }
+    }
+
+    public Task WaitAsync()
+    {
+        lock (_mutex)
+        {
+            return _tcs.Task;
+        }
+    }
+
+    public Task WaitAsync(int millisecondTimeout)
+    {
+        var waitTask = WaitAsync();
+
+        return waitTask.IsCompleted ? waitTask : InternalWaitAsync(waitTask, millisecondTimeout);
```

Comment:
if `completedTask != task`, the "usual" behaviour would be to throw a `TimeoutException` or similar right? Currently there's no way to signal to the caller that the task timed out?

Also, this is built-in in .NET 6, so you could do something like:
```suggestion
        return waitTask.IsCompleted 
            ? waitTask 
#if NET6_0_OR_GREATER
            : waitTask.WaitAsync(TimeSpan.FromMilliseconds(millisecondTimeout));
#else            
            : InternalWaitAsync(waitTask, millisecondTimeout);
#endif            
```

Reply:
>if completedTask != task, the "usual" behaviour would be to throw a TimeoutException or similar right?
for a `Task` but not for `ManualResetEventSlim` what I'm going to do is to copy the signature and return a boolean to signal the caller.

https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Threading/ManualResetEventSlim.cs,447

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/3609/commits/18bfc35e3ce644ce86d4f155a7ffde9667c3a1d8